lib/Sema/SemaExpr.cpp

```diff
@@ -12490,36 +12497,66 @@ ExprResult Sema::ActOnBoundsCastExpr(Scope *S, SourceLocation LParenLoc,
     }
   }
 
-  return BuildBoundsCastExpr(LParenLoc, castTInfo, RParenLoc, E1,
-                             dyn_cast<BoundsExpr>(bounds.get()), kind);
+  return BuildBoundsCastExpr(OpLoc, Kind, castTInfo,
+                             SourceRange(LAngleBracketLoc, RAngleBracketLoc),
+                             SourceRange(LParenLoc, RParenLoc), E1,
+                             dyn_cast<BoundsExpr>(bounds.get()));
 }
 
-ExprResult Sema::GenerateBoundsExpr(Expr *E2, Expr *E3) {
+ExprResult Sema::GenerateBoundsExpr(Expr *E1, Expr *E2, Expr *E3,
+                                    QualType DestTy, int SyntaxType,
+                                    tok::TokenKind Kind,
+                                    SourceLocation TypeLoc) {
   ExprResult Result(true);
   ExprResult ResE2 = CorrectDelayedTyposInExpr(E2);
   ExprResult ResE3 = CorrectDelayedTyposInExpr(E3);
 
-  if (!ResE2.isUsable() && !ResE3.isUsable()) {
-    llvm::APInt I = llvm::APInt(1, 1, false);
-    uint64_t Bits = I.getZExtValue();
-    unsigned Width = Context.getIntWidth(Context.UnsignedLongLongTy);
-    llvm::APInt ResultVal(Width, Bits);
-    IntegerLiteral *One = IntegerLiteral::Create(
-        Context, ResultVal, Context.UnsignedLongLongTy, SourceLocation());
-    Result =
-        ActOnCountBoundsExpr(SourceLocation(), BoundsExpr::Kind::ElementCount,
-                             One, SourceLocation());
-  } else if (ResE2.isUsable() && !ResE3.isUsable()) {
-    Result =
-        ActOnCountBoundsExpr(SourceLocation(), BoundsExpr::Kind::ElementCount,
-                             ResE2.get(), SourceLocation());
-  } else if (ResE2.isUsable() && ResE3.isUsable()) {
-    Result = ActOnRangeBoundsExpr(SourceLocation(), ResE2.get(), ResE3.get(),
-                                  SourceLocation());
+  switch (SyntaxType) {
+  case BoundsCastExpr::SyntaxType::Single:
+    if (DestTy->isCheckedPointerPtrType() || DestTy->isUncheckedPointerType()) {
+      llvm::APInt I = llvm::APInt(1, 1, false);
+      uint64_t Bits = I.getZExtValue();
+      unsigned Width = Context.getIntWidth(Context.UnsignedLongLongTy);
+      llvm::APInt ResultVal(Width, Bits);
+      IntegerLiteral *One = IntegerLiteral::Create(
+          Context, ResultVal, Context.UnsignedLongLongTy, SourceLocation());
+      Result =
+          ActOnCountBoundsExpr(SourceLocation(), BoundsExpr::Kind::ElementCount,
+                               One, SourceLocation());
+    } else {
+      Diag(TypeLoc, diag::err_bounds_cast_error_with_single_syntax);
+    }
+    break;
+  case BoundsCastExpr::SyntaxType::Count:
+    if (!DestTy->isCheckedPointerArrayType()) {
+      Diag(TypeLoc, diag::err_bounds_cast_error_with_array_syntax);
+      Result = CreateInvalidBoundsExpr();
+    } else if ((E1->getType())->isVoidPointerType()) {
+      Diag(E1->getLocStart(),
+             diag::err_typecheck_void_pointer_count_return_bounds);
+    } else {
+      if (ResE2.isUsable())
```

Comment:
I thought E2, E3 and E1 are not invalid after they comes here. I will add invalid check.